proposals/0000-existentials.rst

```diff
@@ -0,0 +1,587 @@
+      type to become ``ty`` and assume ``C`` as a given, available for use anywhere in the innermost
+      case-match or lambda, including invisible lambdas as introduced in the argument to a function
+      with a higher-rank type. (This is rule ``IGiven`` in Fig. 11 of the paper_.) Use of such a given
+      may have a surprising influence on runtime behavior, see Ambiguity_, below.
+
+   #. When inferring the type of a lambda-expression ``\ pats -> e``, we must ensure that no variable bound
+      in ``pats`` leaks into the type of the expression; this is like a skolem-escape.
+
+      (More technically) When inferring the type of a lambda-expression ``\ pats -> e``: We must bump the
+      type-checker level before checking ``e``. This is because variables bound in ``pats`` might be mentioned
+      in ``Witness``\ es in the type of ``e``. Then, when forming the type of the whole lambda-expression,
+      we promote the inferred type of ``e``; if any variables bound in ``pats`` are encountered, issue an
+      error.
+
+   #. When trying to satisfy a class constraint ``[W] C (exists (a :: k). ty)``, instead solve
+      ``[W] forall (a :: k). C ty``. For multi-parameter type classes, apply this treatment one
+      argument at a time. Note that this treatment applies to ``~``, which is solved like any
+      other type class (recall that lifted ``~`` is distinct from unlifted ``~#``, to which this
+      treatment does not apply).
+
+   #. When trying to satisfy a class constraint ``[W] C (C' /\ ty)``, instead solve
+      ``[W] C' => C ty``. That is, assume ``C'`` as a given and then solve ``[W] C ty``.
+      For multi-parameter type classes, apply this treatment one argument at a time.
+
+#. ``GHC.Exists`` exports ::
+
+     discardEvidence :: (() => c /\ a) -> a
+     discardEvidence x = x
+
+   This function effectively prevents the evidence for ``c`` from being used. Note that
+   it has a (trivially) higher-rank type, so that GHC introduces an invisible lambda around
+   its argument. Without this invisible lambda, the given would "escape" and be usable.
+
+   The ``discardEvidence`` function itself is completely ordinary and could be defined by
+   users, but the small unexpected twist in its type (and its general usefulness) suggests
+   it should be ``GHC.Exists``.
+
+#. **Core language.** There are several modifications to the Core language necessary to
+   support this proposal. The notes here echo the design in the paper_, Section 5.
+
+   1. The ``exists`` type would need to be added to Core as a new constructor of ``Type``::
+
+        | ExistsTy TyVar Type
+          -- typing rule:
+          -- Γ ⊢ ki : Type
+          -- Γ, tv:ki ⊢ inner_ty : TYPE rep
+          -- tv # rep
+          -- ------------------------------------
+          -- Γ ⊢ ExistsTy (tv:ki) inner_ty : TYPE rep
+
+   #. The ``Witness`` type would need to be added to Core as a new constructor of ``Type``::
```

Comment:
Don't we currently use the same `Type` definition in source and core? Or is that no longer the case?

Reply:
I thought they are currently the same, but the hope is to make them different!

Reply:
I guess it's an implementation detail that doesn't have to worry us here.

Reply:
I don't really understand the question. We use `HsType` in source and `Type` in core. But we also desugar the `HsType` to a `Type` before doing any type comparisons. So maybe that's what you mean by using `Type` in source?

Maybe this is the missing piece: ``Witness``, though it looks like a type constructor, isn't. That is, we can't make a `TyCon` spelled `Witness` and use `TyConApp` to represent applications of it. Instead, it needs to be its own custom form.

Reply:
I was under the impression that `HsType` is for “types expressions as written in the source code”, but the Source type checker uses `Type` to describe the types of functions, variables etc. But probably that doesn’t matter here, feel free to resolve this thread.

Reply:
Further thought revealed I had overlooked something here. Fixed in edits I'm making now.